### ISSUE TITLE
Nested unions are "untagged"

### ISSUE THREAD

**Issue by OliverJAsh** [pelotom/unionize#61]
I just ran into this bug which resulted in a runtime exception. 😢

``` ts
import { unionize, ofType, UnionOf } from 'unionize';

const NestedAction = unionize({ Foo: {} }, { value: 'value' });
type NestedAction = UnionOf<typeof NestedAction>;

const Action = unionize({ Nested: ofType<NestedAction>() }, { value: 'value' });

declare const wrong: { value: {} };

// No error!!!
Action.Nested(wrong);

// Expected: (value: { tag: "Foo"; } & { value: {}; }) => …
// Actual: (value: Pick<{ tag: "Foo"; } & { value: {}; }, "value">) => …
// Why is `tag` omitted??
Action.Nested;
```

One known workaround is to force the nested union tag to be different:

``` diff
-const NestedAction = unionize({ Foo: {} }, { value: 'value' });
+const NestedAction = unionize({ Foo: {} }, { tag: 'tag2', value: 'value' });
```

However that's a non-option when modelling Redux actions, since the tag must be `type` to conform to the usual Redux action type.

**Comment by OliverJAsh** [comment]
This change appears to fix it, but I'm probably breaking some other behaviour. Not sure what exactly because it seems there isn't a test for it.

``` diff
 export declare type Creators<Record, TaggedRecord, TagProp extends string> = {
-  [T in keyof Record]: {} extends Required<UnTagged<Record[T], TagProp>>
+  [T in keyof Record]: {} extends Required<Record[T]>
     ? ((value?: {}) => TaggedRecord[keyof TaggedRecord])
-    : ((value: UnTagged<Record[T], TagProp>) => TaggedRecord[keyof TaggedRecord])
+    : ((value: Record[T]) => TaggedRecord[keyof TaggedRecord])
```

``` ts
// error :-)
Action.Nested(wrong);
// no error :-)
Action.Nested(NestedAction.Foo({}));
```

**Comment by OliverJAsh** [comment]
It seems `UnTagged` was added to allow the behaviour shown in this test:

https://github.com/sledorze/unionize/blob/9763034266f9b5aae61fcd93ecf8ea70efe7ad11/src/index.spec.ts#L94-L105

https://github.com/pelotom/unionize/pull/36/files#diff-6b623d6bcf5e7f06c466aa060ec9c4b6R94

However this test was removed recently: https://github.com/pelotom/unionize/commit/87456d3

Furthermore, I just tried to restore the test and it seems to be throwing a type error.

So, where does this leave us with the issue I described originally?

Either:

- We still need `UnTagged` to allow for the behaviour shown in the test. If that's the case, we should 1) revive that test and 2) fix it because that test is now failing.
- We no longer need `UnTagged`, as we don't need the behaviour shown in the test. If that's the case, removing it will fix this issue.

@pelotom WDYT?
